### DISCUSSION TITLE
Dudas TPI

### DISCUSSION THREAD

**Discussion by Gdepablo** [dds-jv/foro#388] [Consultas de persistencia] [answered]
Buenas, estoy teniendo un error al tirar el contextUp del TPI porque me dice que EntityManager isNull. Probé cambiar el persistance con el del tp y eso, pero nada, sigue dando igual, y no entiendo por qué.
Además, me surgió una duda, ¿cuándo se usa el @MappedSuperclass? También @Table cuándo corresponde que se use? Porque veo que en estrategias single table -según me dijo franco- se usa el discriminator value (aunque no me termina de cerrar por qué, entiendo que es como para que el framework sepa que esa tabla / lo que sea es una entidad con herencia?

Después el tema -especificamente- del TPI de Juego de Tronos, donde regiones y lugares es many to many, no me cierra. Una región no debería tener muchos lugares, pero un lugar una única región? 

Por otro lado, el tema JOINED vs SINGLE TABLE, cuál sería la gran diferencia ? Entiendo que Single table te pone todo en una tabla, pero joined te une dos tablas? (Así es como lo entiendo pensándolo en SQL)

**Comment by gastonprieto** [comment, marked as answer]
Buenas @Gdepablo 

> Buenas, estoy teniendo un error al tirar el contextUp del TPI porque me dice que EntityManager isNull. Probé cambiar el persistance con el del tp y eso, pero nada, sigue dando igual, y no entiendo por qué.

El error de null en general es porque el test que valida que el pedido del entitymanager no de null, el entitymanager devuelve null en el caso de no poder crearlo. Pero fijate que seguramente en el stack trace podes ver el error general.

>Además, me surgió una duda, ¿cuándo se usa el @MappedSuperclass? 

Se usa cuando tenes una jerarquia de herencia y no queres realizar consultas polimorficas en dicha jerarquia (nadie tiene una asociación a esa jerarquía o no se realiza la consulta por la superclase). En general se usa para reutilizar estructura de una superclase , lo que le indica a hibernate es que lea los mapeos solamente.

> También https://github.com/table cuándo corresponde que se use? Porque veo que en estrategias single table -según me dijo franco- se usa el discriminator value (aunque no me termina de cerrar por qué, entiendo que es como para que el framework sepa que esa tabla / lo que sea es una entidad con herencia?

`@table` no tiene que ver con lo de discriminator column ni con herencia. Es para decidirle al ORM el nombre de la tabla. 
El valor del discriminador en una fila de una tabla que usa single table sirve para decirle al framework que tipo de objeto es porque sino no sabe como distinguir las filas de la single table a que tipo de la jerarquia de herencia corresponde

> Después el tema -especificamente- del TPI de Juego de Tronos, donde regiones y lugares es many to many, no me cierra. Una región no debería tener muchos lugares, pero un lugar una única región?

Fijate que en la solución dice que las regiones se pueden solapar, entonces un lugar podria estar en uno de esos solapamientos

> Por otro lado, el tema JOINED vs SINGLE TABLE, cuál sería la gran diferencia ? Entiendo que Single table te pone todo en una tabla, pero joined te une dos tablas? (Así es como lo entiendo pensándolo en SQL)

Correcto Single table, joined no te une dos tablas pero para formar un objeto completo. pero la cantidad de tablas depende de la cantidad de subclases que haya, 1 + n . Esa es una gran diferencia, lo mismo que no necesitas discriminador
